packages/zoltan2/src/algorithms/partition/Zoltan2_AlgMultiJagged.hpp

```diff
@@ -78,7013 +73,9090 @@
 #endif
 #endif
 
-#ifdef HAVE_ZOLTAN2_OMP
-#include <omp.h>
+// This macro could just be KOKKOS_HAVE_CUDA but preserving it here to
+// facilitate testing this on and off. When ZOLTAN2_MJ_USE_CUDA_KERNEL is
+// defined, the main kernel is run using a parallel_for -> parallel_for loop
+// with atomics, instead of parallel_reduce -> parallel_reduce which gives
+// better performance on host. This occurs because CUDA is more efficient with
+// atomics than host code. I understand ScatterView may be a mechanism to allow
+// a single code pipe-line which runs both reduction or atomic patterns but did
+// not investigate that yet.
+// TODO: Implement ScatterView and get rid of this macro.
+#ifdef KOKKOS_HAVE_CUDA
+#define ZOLTAN2_MJ_USE_CUDA_KERNEL // Atomic Atomic Loops
 #endif
 
+// TODO: This macro will be removed after some further profiling.
+// This option is being maintained to evaluate the perfomance of using floats
+// versus doubles for the reduction arrays in the main kernel of MJ.
+#define ZOLTAN2_MJ_USE_FLOAT_ARRAY_FOR_KERNEL
+
 #define LEAST_SIGNIFICANCE 0.0001
 #define SIGNIFICANCE_MUL 1000
 
 //if the (last dimension reduce all count) x the mpi world size
 //estimated to be bigger than this number then migration will be forced
 //in earlier iterations.
 #define FUTURE_REDUCEALL_CUTOFF 1500000
+
 //if parts right before last dimension are estimated to have less than
 //MIN_WORK_LAST_DIM many coords, migration will be forced in earlier iterations.
 #define MIN_WORK_LAST_DIM 1000
 
-
-
-
 #define ZOLTAN2_ABS(x) ((x) >= 0 ? (x) : -(x))
+
 //imbalance calculation. Wreal / Wexpected - 1
 #define imbalanceOf(Wachieved, totalW, expectedRatio) \
```

Comment:
See note about non-namespaced macros.

Reply:
Changed all macros to be ZOLTAN2 prefixed.